### DISCUSSION TITLE
One-shot boot target / UEFI variables

### DISCUSSION THREAD

**Discussion by rowlap** [ipxe/ipxe#1309] [Ideas]
I'm interested in having a server boot to a non-standard target, but only once, via iPXE.

Options include:
- Modify the config file in-between reboots
- Call out to some API which only works once
- Provide a cut-off time
- Modify a UEFI variable on the server, from iPXE

The difficulty with option 1 is knowing when iPXE has consumed the one-off config.

Option 2 is feasible, if we provide a central coordinator, except the API is limited to DNS lookups, e.g.
```
#!ipxe

nslookup ${net0/hostname}.oneshot.example.com && goto oneshot
```

Option 3 could make use of `unixtime`, except we can't currently test for inequalities.

Hence my pitch for Option 4; if iPXE could delete UEFI variables, it would be possible to pre-set the oneshot behaviour, then have iPXE undo that change. Mapping to existing iPXE verbs, deletion could use `clear`. (Why am I not pitching the C or U of CRUD methods? It avoids having to discuss the GUID problem.)

**Comment by mcb30** [comment]
Your option 2 is not limited to DNS lookups.  Provide an HTTP endpoint that examines its query parameters and returns an iPXE script containing the commands that iPXE should execute next.  Call this HTTP endpoint using `chain` from within iPXE.  For example:

```
#!ipxe
chain http://webserver/bootnext?mac=${netX/mac}&uuid=${uuid}
```
with the `bootnext` endpoint returning an iPXE script directing the next stage of the boot, such as
```
#!ipxe
chain http://webserver/fedora/boot.ipxe
```

If you want to separate out policy from mechanism, you can have the HTTP endpoint return an iPXE script that just sets a variable, and then consume this variable from within your original script.  (You may want to use `chain --autofree ...` so that the endpoint's script is automatically freed after execution, to simplify your code.)

Once you've got that part working, you can approach your original problem.  Configure your "one-shot" state on the HTTP server side, and make it sticky.  The "one-shot" boot state should probably be cleared only after you have some indication of a **successful** boot into this one-shot option.  This would mean having the one-shot booted OS take some action to trigger the state being cleared.

This gives you the ability to make sure that the one-shot boot actually **succeeds** once, rather than merely being **attempted** once.  It will be robust against transient errors, and you can continue to rely on safeguards such as the UEFI watchdog timer (which will reset the system after several minutes of unsuccessful boot attempts).

**Comment by rowlap** [comment]
Thanks for the suggestions! I was leaning towards a single attempt, rather than a successful boot, to avoid the risk of boot loops, but this is policy which could be implemented either way.

Setting a variable is ideal as it fails-safe, should the HTTP endpoint somehow become unreachable. It's feasible to implement some "clear on read" semantics for the endpoint.

There are some security issues to think through, although the worst-case-scenario might be the nuisance effect of clearing the oneshot state. I'm not sure I've made a compelling case for UEFI variable deletion, so will continue as discussed for the moment.
